### ISSUE TITLE
rsyslog size maxed out at 1mb, does not scale to large logs

### ISSUE THREAD

**Issue by patpatpat123** [rsyslog/rsyslog#5195]
Hello team,

First post in this repo. If not anything else, a big thank you for this cool project.
Our business use case is simple. we have logs, which keep growing, sort of time series.
However, one single entry of the logs, is quite large, sometimes up to 5mb, for one single entry.
Using rsyslog, we then ship those entries to our message bus, for further downstream processing.

### Expected behavior
I would expect rsyslog to be able to handle those logs, even the entries that are big.

### Actual behavior
Maxed out at around 1mb, rsyslog truncates the log. Therefore, if we have a 3 Mb, 5mb log, rsyslog is not able to ship the original log.

### Steps to reproduce the behavior
Any logline>1mb would prove this finding.

Would it be possible to have rsyslog ship large single line entries, please?

**Comment by davidelang** [comment]
did you configure maxmessagesize? what version are you running?

David Lang

**Comment by patpatpat123** [comment]
Hello David,

Thank you for looking into this:

We are currently using rsyslog version 8.2112.

And the property is set as follow:``` $MaxMessageSize 524288```
We tried bumping the number to above 1 mb as well, still issue

Could you please provide some advices on this issue?

**Comment by rgerhards** [comment]
@patpatpat123 pls share your complete config

**Comment by patpatpat123** [comment]
Hello @rgerhards 

Thank you for looking into this.

Here we go:

```
action(type="omkafka"
           broker={{ brokers }}
           partitions.auto="on"
           topic="my_topic"
           Template="MyKafkaMessage"
           resubmitOnFailure="on"
           KeepFailedMessages="on"
           failedMsgFile="/var/log/rsyslog/failed_kafka_msg.log"
           confParam=[
             "compression.codec=snappy",
             "security.protocol=ssl",
             "ssl.certificate.location=/etc/rsyslog.d/ssl/a.crt",
             "ssl.key.location=/etc/rsyslog.d/ssl/a.key",
             "ssl.ca.location=/etc/rsyslog.d/ssl/ca.crt"
           ]
    )
```

**Comment by patpatpat123** [comment]
```
$MaxMessageSize 524288

{% if 'logstash' in grains['id'] -%}
module(load="imjournal" IgnorePreviousMessages="on")
{% endif -%}

module(load="imuxsock")
module(load="imudp")
input(type="imudp" port="514")

module(load="imtcp")
input(type="imtcp" port="514")

#### GLOBAL DIRECTIVES ####
# Use FQDN instead of short names
$PreserveFQDN on

$IncludeConfig /etc/rsyslog.d/*.conf

cron.*                                                  /var/log/cron

```

**Comment by patpatpat123** [comment]
```
module(load="impstats"
       interval="60"
       log.syslog="on"
       resetcounters="on"
       severity="7")

$MaxMessageSize 524288

# Provides support for local system logging
module(load="imuxsock")
# Provides UDP syslog reception
module(load="imudp") # needs to be done just once
input(type="imudp" port="514")

# Provides TCP syslog reception
module(load="imtcp") # needs to be done just once
input(type="imtcp" port="514")

module(load="omkafka")

#### GLOBAL DIRECTIVES ####
# Use FQDN instead of short names
$PreserveFQDN on

global (
     parser.escapeControlCharacterTab="off"
   )

$IncludeConfig /etc/rsyslog.d/*.conf

$template RemoteHost,"/srv/rsyslog/%programname%/%$year%-%$month%-%$day%/%HOSTNAME%.log"
$template RemoteHostAudit,"/srv/rsyslog/%programname%/%$year%-%$month%-%$day%/audit_%HOSTNAME%.log"

if $fromhost == "{{ grains.id }}" then {
syslog.=debug /var/log/rsyslog-stats

cron.*                                                  /var/log/cron

authpriv.* ?RemoteHost
} else if ($msg contains "audit_category" and ($msg contains "AUTHENTICATED" or $msg contains "FAILED_LOGIN")) then {
  *.* ?RemoteHostAudit
} else {
  *.* ?RemoteHost
}


```

**Comment by patpatpat123** [comment]
are those enough?

**Comment by rgerhards** [comment]
It's hard to grasp the sequence and thus what applies when. Maybe it's best if you create a startup debug log. Enable debug logging, let rsyslog start and run for a few seconds and then shut it down again.

Instructions: https://www.rsyslog.com/doc/v8-stable/troubleshooting/howtodebug.html

Once this is done, please post the debug log (e.g. via pastbin or github). After that, we can review the contents and check which configuration rsyslog actually uses (well, mostly). 

Note: that may gives no sufficient clue. In that case we would need to look at a debug log where the problem in question actually happens. That should give ultimate insight (but is the hardest to capture and work with).

**Comment by patpatpat123** [comment]
Just a quick question if I may.
Does rsyslog limits the message size up to a certain size?

My expectation is that if I put a crazy value to ```$MaxMessageSize```, let's say 42G, just for the sake of the example.
Will rsyslog honor, and try to forward a log of size up to 42G, or does it has any internal limitation which will limit to another value?

Thank you

**Comment by rgerhards** [comment]
Sorry for the very late reply. This may no longer be directly relevant to the original deployment, but I am adding the answer for people who find this issue through search while we work through the backlog. We do not lightly close old issues and we do not close them just because of age.

There is no fixed rsyslog 1 MiB message-size ceiling. The first limit to check is rsyslog's global `maxMessageSize` setting. In the config shown above, it is:

```rsyslog
$MaxMessageSize 524288
```

That is 512 KiB, so truncation above that size would be expected. For current-style config, set it high enough before the relevant inputs are configured, for example:

```rsyslog
global(maxMessageSize="6m")
```

For multi-MiB single messages, also check the full path, not only rsyslog:

- UDP is a poor fit for very large single messages because datagram size, fragmentation, and loss behavior become limiting. TCP or RELP is usually the better choice.
- Stream inputs need compatible framing from the sender.
- `omkafka` also depends on librdkafka, broker, and topic message-size settings. Kafka-side limits such as producer and broker maximum message bytes can truncate or reject large records independently of rsyslog.

The earlier maintainer note in this issue also confirms that rsyslog itself is used with much larger messages, for example around 20 MiB in some support cases. I am closing this as answered/configuration guidance. If the problem persists on a current rsyslog version after `maxMessageSize` and downstream Kafka limits are adjusted, please open a new issue with a minimal reproducer, current rsyslog version, full sanitized config, transport used, and the exact truncation/error evidence.
